source/Mlos.Python/mlos/unit_tests/TestSmartCacheWithRemoteOptimizerV3.py

```diff
@@ -36,19 +36,29 @@ def setup_method(self, method):
         self.logger = create_logger('TestSmartCacheWithRemoteOptimizer')
         self.logger.level = logging.DEBUG
 
-        # Start up the gRPC service.
+        # Start up the gRPC service. Try a bunch of times before giving up.
         #
-        self.server = OptimizerMicroserviceServer(port=50051, num_threads=10)
-        self.server.start()
-
-        self.optimizer_service_grpc_channel = grpc.insecure_channel('localhost:50051')
-        self.bayesian_optimizer_factory = BayesianOptimizerFactory(grpc_channel=self.optimizer_service_grpc_channel, logger=self.logger)
+        max_num_tries = 100
+        num_tries = 0
+        for port in range(50051, 50051 + max_num_tries):

```

Comment:
Hmm, questions ...
Is this only necessary when running tests in parallel?
Or is it not cleaning up after itself?  Or maybe some delay after the teardown is necessary?
Can a single remote optimizer not handle requests from multiple clients?

Reply:
In fd8ef0884dc6aa9c0a03144f1027520683557431 I think we had to add some additional code to the `teardown_method`:
https://github.com/microsoft/MLOS/pull/154/commits/e29697b82de38418ab7ab7b8d1da456037e50c9e

Reply:
It's so that multiple services can run in parallel. Having one handle multiple requests would work too.

Reply:
*one service - each test instantiates its own optimizer.

Reply:
Thanks, makes sense.
The teardown changes (specifically closing the client channel explicitly) may still be necessary in the other one.  I think I only made the change in the one that was throwing errors at the time.

Reply:
Hmmm I thought that the client could just crash... what happens if we don't close the connection on the client?

Reply:
If I recall correctly it leaves a socket open that causes conflicts later on when it goes to reuse it.

Reply:
If I recall correctly it leaves a socket open that causes conflicts later on when it goes to reuse it.

Reply:
I thought it would only happen on the server side. Is there any repro that would tell us when we've fixed the bug?

Reply:
https://github.com/microsoft/MLOS/compare/main...bpkroth:demonstrate-missing-grpc-client-close-bug
https://github.com/bpkroth/MLOS/commit/cadc29c8b50cf110b3325c36431cc92e2d2c1189

Reply:
awesome, thanks